cmd/copy/main.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/containerd/continuity/fs"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/pkg/errors"
+	"github.com/tonistiigi/copy/detect"
+	"github.com/tonistiigi/copy/user"
+	copy "github.com/tonistiigi/fsutil/copy"
+)
+
+// cp with Dockerfile ADD/COPY semantics
+
+type opts struct {
+	unpack bool
+	chown  *copy.ChownOpt
+	root   string
+}
+
+type chown struct {
+	uid, gid int
+}
+
+func main() {
+	var opt opts
+	var username string
+	flag.BoolVar(&opt.unpack, "unpack", false, "")
+	flag.StringVar(&username, "chown", "", "")
+
+	flag.Parse()
+	args := flag.Args()
+
+	if username != "" {
+		uid, gid, err := user.GetUser(appcontext.Context(), "/", username)
+		if err != nil {
+			panic(err)
+		}
+		opt.chown = &copy.ChownOpt{Uid: int(uid), Gid: int(gid)}
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+	opt.root = wd
+
+	if err := runCopy(appcontext.Context(), args, opt); err != nil {
+		panic(err)
+	}
+}
+
+func runCopy(ctx context.Context, args []string, opt opts) error {
+	if len(args) < 2 {
+		return fmt.Errorf("invalid args %v", args)
+	}
+
+	srcs := args[:len(args)-1]
+	isdir := false
+
+	for i, src := range srcs {
+		fi, err := os.Lstat(src)
+		if err == nil && fi.IsDir() {
+			isdir = true
+			srcs[i] = path.Clean(src) + "/."
+		}
+	}
+
+	if len(srcs) > 1 {
+		isdir = true
+
+	}
+
+	dest := args[len(args)-1]
+
+	// This handles the case where destination is a symlink.
+	if fi, err := os.Lstat(path.Clean(dest)); err == nil {
```

Comment:
@tonistiigi I used `path.Clean` instead of `filepath.Clean`, let me know if you want this changed back to filepath (which would be a new dependency)